File: ndi-python/ndi/cloud/sync/download_new.py
"""Download new documents (and associated data files) from remote to local.

Ported from: ndi.cloud.sync.downloadNew (MATLAB)
"""

from typing import TYPE_CHECKING

from ndi.cloud.sync.sync_options import SyncOptions
from ndi.cloud.sync.internal.list_local_documents import list_local_documents
from ndi.cloud.sync.internal.list_remote_document_ids import list_remote_document_ids
from ndi.cloud.sync.internal.download_ndi_documents import download_ndi_documents
from ndi.cloud.sync.internal.index.read_sync_index import read_sync_index
from ndi.cloud.sync.internal.index.update_sync_index import update_sync_index

if TYPE_CHECKING:
    from ndi.dataset import Dataset


def download_new(ndi_dataset: 'Dataset', **kwargs) -> None:
    """Download new documents (and associated data files) from remote to local.

    This function implements the "DownloadNew" synchronization mode.
    It identifies documents present on the remote cloud storage that are
    not present in the local NDI dataset and downloads them.

    No local documents are deleted or modified.
    No remote documents are deleted or modified.

    It relies on a sync index file ([NDIDATASET.path]/.ndi/sync/index.json)
    to keep track of previously synced states, though for "DownloadNew"
    mode, it compares the current remote state to the remote state from the
    last sync (from the index) to identify newly added remote documents.
    The index is updated after the operation.

    Args:
        ndi_dataset: The local NDI dataset object
        **kwargs: Optional synchronization options:
            - sync_files (bool): If True, files will be synced (default: True)
            - verbose (bool): If True, verbose output is printed (default: True)
            - dry_run (bool): If True, actions are simulated but not performed (default: False)
            - file_upload_strategy (str): "serial" or "batch" (default: "batch")

    See also:
        ndi.cloud.sync_dataset
        ndi.cloud.sync.SyncOptions
        ndi.cloud.sync.SyncMode
    """
    from ndi.cloud.internal.get_cloud_dataset_id_for_local_dataset import get_cloud_dataset_id_for_local_dataset

    sync_options = SyncOptions(**kwargs)

    if sync_options.verbose:
        print(f'Syncing dataset "{ndi_dataset.path}".')
        print('Will download new documents from remote.')

    # Resolve cloud dataset identifier
    cloud_dataset_id = get_cloud_dataset_id_for_local_dataset(ndi_dataset)
    if sync_options.verbose:
        print(f'Using Cloud Dataset ID: {cloud_dataset_id}')

    # 1. Read sync index
    sync_index = read_sync_index(ndi_dataset)
    if sync_index is None or not sync_index.get('remoteDocumentIdsLastSync'):
        remote_ids_last_sync = []
    else:
        remote_ids_last_sync = sync_index['remoteDocumentIdsLastSync']

    if sync_options.verbose:
        print(f'Read sync index. Last sync recorded {len(remote_ids_last_sync)} remote documents.')

    # 2. Get current remote state - Returns dict with ndi_id, api_id
    remote_document_id_map = list_remote_document_ids(cloud_dataset_id)
    current_remote_ndi_ids = remote_document_id_map['ndi_id']

    # 3. Calculate differences: documents added to remote since last sync
    ndi_ids_to_download = []
    cloud_api_ids_to_download = []
    remote_ids_last_sync_set = set(remote_ids_last_sync)
    for i, ndi_id in enumerate(current_remote_ndi_ids):
        if ndi_id not in remote_ids_last_sync_set:
            ndi_ids_to_download.append(ndi_id)
            cloud_api_ids_to_download.append(remote_document_id_map['api_id'][i])

    if sync_options.verbose:
        print(f'Found {len(ndi_ids_to_download)} documents added on remote since last sync.')

    # 4. Perform download actions
    if ndi_ids_to_download:
        if sync_options.dry_run:
            print(f'[DryRun] Would download {len(ndi_ids_to_download)} documents from remote.')
            if sync_options.verbose:
                for i, ndi_id in enumerate(ndi_ids_to_download):
                    print(f'  [DryRun] - NDI ID: {ndi_id} '
                          f'(Cloud Specific ID: {cloud_api_ids_to_download[i]})')
        else:
            if sync_options.verbose:
                print(f'Downloading {len(ndi_ids_to_download)} documents...')

            # This internal function handles batch download and adding to ndiDataset,
            # respecting sync_options.sync_files and sync_options.verbose internally.
            download_ndi_documents(
                cloud_dataset_id, cloud_api_ids_to_download, ndi_dataset, sync_options
            )

            if sync_options.verbose:
                print(f'Completed downloading {len(ndi_ids_to_download)} documents.')
    else:
        if sync_options.verbose:
            print('No new documents to download from remote.')

    # 5. Update sync index
    if not sync_options.dry_run:
        # Update local state after download
        _, final_local_document_ids = list_local_documents(ndi_dataset)

        update_sync_index(
            ndi_dataset, cloud_dataset_id,
            local_document_ids=final_local_document_ids,
            remote_document_ids=remote_document_id_map['ndi_id']
        )

        if sync_options.verbose:
            print('Sync index updated.')

    if sync_options.verbose:
        print(f'Syncing complete for dataset: {ndi_dataset.path}')

File: src/ndi/+ndi/+cloud/+sync/downloadNew.m
function [success, errorMessage, report] = downloadNew(ndiDataset, syncOptions)
%DOWNLOADNEW Download new documents (and associated data files) from remote to local.
%
% Syntax:
%   [SUCCESS, ERRORMESSAGE, REPORT] = ndi.cloud.sync.downloadNew(NDIDATASET, SYNCOPTIONS)
%
%   This function implements the "DownloadNew" synchronization mode.
%   It identifies documents present on the remote cloud storage that are
%   not present in the local NDI dataset and downloads them.
%
%   No local documents are deleted or modified.
%   No remote documents are deleted or modified.
%
%   It relies on a sync index file ([NDIDATASET.path]/.ndi/sync/index.json)
%   to keep track of previously synced states, though for "DownloadNew"
%   mode, it compares the current remote state to the remote state from the
%   last sync (from the index) to identify newly added remote documents.
%   The index is updated after the operation.
%
%   Inputs:
%       ndiDataset (1,1) ndi.dataset - The local NDI dataset object.
%       syncOptions (name, value pairs) - Optional synchronization options:
%       - SyncFiles (logical) - If true, files will be synced (default: false).
%       - Verbose (logical) - If true, verbose output is printed (default: true).
%       - DryRun (logical) - If true, actions are simulated but not performed (default: false).
%
%   Outputs:
%       success (logical) - True if the operation completed successfully, false otherwise.
%       errorMessage (string) - Error message if success is false, empty otherwise.
%       report (struct) - Structure containing details of the changes applied.
%           Fields:
%           - downloaded_document_ids (string array)
%           - failed (string array) - NDI IDs that were requested from the
%             remote but did not arrive. These are deliberately left out of
%             the sync index so the next sync requests them again.
%
%   See also:
%       ndi.cloud.syncDataset
%       ndi.cloud.sync.SyncOptions,
%       ndi.cloud.sync.enum.SyncMode

    arguments
        ndiDataset (1,1) ndi.dataset
        syncOptions.?ndi.cloud.sync.SyncOptions
    end

    success = true;
    errorMessage = '';
    report = struct('downloaded_document_ids', string.empty, ...
        'failed', string.empty);

    % Remote documents we asked for and did not get. Declared here so the
    % index update in step 5 can see it whichever branch below ran.
    notObtainedNdiIds = string.empty;

    try
        syncOptions = ndi.cloud.sync.SyncOptions(syncOptions);

        if syncOptions.Verbose
            fprintf(['Syncing dataset "%s". \nWill download new documents ', ...
                'from remote.\n'], ndiDataset.path);
        end

        % Resolve cloud dataset identifier
        cloudDatasetId = ndi.cloud.internal.getCloudDatasetIdForLocalDataset(ndiDataset);
        if syncOptions.Verbose
            fprintf('Using Cloud Dataset ID: %s\n', cloudDatasetId);
        end

        % Wait for any in-flight bulk file extractions before inventorying
        % so the remote document/file list we're about to read is stable.
        if ~syncOptions.DryRun
            if syncOptions.Verbose
                fprintf('Waiting for any in-flight bulk file uploads to finish before inventorying...\n');
            end
            ndi.cloud.api.files.waitForAllBulkUploads(cloudDatasetId);
        end

        % 1. Read sync index
        syncIndex = ndi.cloud.sync.internal.index.readSyncIndex(ndiDataset);
        if isempty(syncIndex) || isempty(syncIndex.remoteDocumentIdsLastSync)
            remoteIdsLastSync = strings(0,1); % Ensure it's an empty string array for setdiff
        else
            remoteIdsLastSync = string(syncIndex.remoteDocumentIdsLastSync); % Ensure string array
        end
        if syncOptions.Verbose
            fprintf('Read sync index. Last sync recorded %d remote documents.\n', numel(remoteIdsLastSync));
        end

        % 2. Get current remote state - Returns Nx2 table with variables ndiId, apiId
        remoteDocumentIdMap = ndi.cloud.sync.internal.listRemoteDocumentIds(cloudDatasetId);
        currentRemoteNdiIds = remoteDocumentIdMap.ndiId;

        % 3. Calculate differences: documents added to remote since last sync
        [ndiIdsToDownload, indicesToDownload] = setdiff(currentRemoteNdiIds, remoteIdsLastSync, 'stable');

        if syncOptions.Verbose
            fprintf('Found %d documents added on remote since last sync.\n', ...
                numel(ndiIdsToDownload));
        end

        % 4. Perform download actions
        if ~isempty(ndiIdsToDownload)
            cloudApiIdsToDownload = remoteDocumentIdMap.apiId(indicesToDownload);

            if syncOptions.DryRun
                fprintf('[DryRun] Would download %d documents from remote.\n', ...
                    numel(ndiIdsToDownload));
                if syncOptions.Verbose
                    for i = 1:numel(ndiIdsToDownload)
                        fprintf('  [DryRun] - NDI ID: %s (Cloud Specific ID: %s)\n', ...
                            ndiIdsToDownload(i), cloudApiIdsToDownload(i));
                    end
                end
            else
                if syncOptions.Verbose
                    fprintf('Downloading %d documents...\n', numel(ndiIdsToDownload));
                end
                % This internal function handles batch download and adding to ndiDataset,
                % respecting syncOptions.SyncFiles and syncOptions.Verbose internally.
                downloadedDocs = ndi.cloud.sync.internal.downloadNdiDocuments(...
                    cloudDatasetId, cloudApiIdsToDownload, ndiDataset, syncOptions);

                if ~isempty(downloadedDocs)
                     % Extract IDs. downloadedDocs is a cell array of ndi.document objects.
                     docIds = cellfun(@(d) d.id(), downloadedDocs, 'UniformOutput', false);
                     report.downloaded_document_ids = string(docIds);
                end

                % What we asked for and did not receive. This has to be
                % recorded: step 3 above works out what to fetch by diffing
                % the remote list against the *index*, not against what is
                % actually local, so any ID written into the index as synced
                % is never requested again.
                notObtainedNdiIds = setdiff(ndiIdsToDownload, ...
                    report.downloaded_document_ids, 'stable');
                report.failed = string(notObtainedNdiIds);

                if ~isempty(notObtainedNdiIds)
                    warning('NDI:cloud:sync:DocumentsNotDownloaded', ...
                        ['%d of %d documents were not downloaded and remain ' ...
                         'only on the remote. They stay outstanding and will ' ...
                         'be retried on the next sync.'], ...
                        numel(notObtainedNdiIds), numel(ndiIdsToDownload));
                end

                if syncOptions.Verbose
                    fprintf('Completed downloading %d of %d documents.\n', ...
                        numel(report.downloaded_document_ids), numel(ndiIdsToDownload));
                end
            end
        else
            if syncOptions.Verbose
                fprintf('No new documents to download from remote.\n');
            end
        end

        % 5. Update sync index
        if ~syncOptions.DryRun

            % Update local state after download
            [~, finalLocalDocumentIds] = ndi.cloud.sync.internal.listLocalDocuments(ndiDataset);

            % Record only the remote documents that actually arrived.
            % Writing the full remote list here was what made a missed
            % document permanent: it lands in remoteDocumentIdsLastSync as
            % though it had synced, and the setdiff in step 3 then excludes
            % it from every future run. (The local side needs no such care --
            % it is re-listed from the dataset above, so it cannot lie.)
            % A filter rather than setdiff: this keeps the list's exact
            % order and shape and removes only the missing entries, where
            % setdiff would also silently deduplicate it.
            syncedRemoteNdiIds = remoteDocumentIdMap.ndiId( ...
                ~ismember(remoteDocumentIdMap.ndiId, notObtainedNdiIds));

            ndi.cloud.sync.internal.index.updateSyncIndex(...
                ndiDataset, cloudDatasetId, ...
                "LocalDocumentIds", finalLocalDocumentIds, ...
                "RemoteDocumentIds", syncedRemoteNdiIds)

            if syncOptions.Verbose
                fprintf('Sync index updated.\n');
            end
        end

        if syncOptions.Verbose
            fprintf('Syncing complete for dataset: %s\n', ndiDataset.path);
        end

    catch ME
        success = false;
        errorMessage = ME.message;
        % Check if syncOptions is an object with Verbose property before accessing
        if exist('syncOptions', 'var') && isprop(syncOptions, 'Verbose') && syncOptions.Verbose
             fprintf('Error in downloadNew: %s\n', errorMessage);
        end
    end
end
